src/blas_like/level3/Gemm/NN_Multistream.hpp

```diff
@@ -0,0 +1,396 @@
+    for (auto id = 0UL; id < num_stream_teams; ++id)
+    {
+        auto const& stream_one = stream_pool.Next();
+        auto const& stream_two = stream_pool.Next();
+
+        auto A1 = A1_MC_STAR.emplace(A1_MC_STAR.end(), g);
+        auto B1 = B1Trans_MR_STAR.emplace(
+            B1Trans_MR_STAR.end(), g);
+        auto C1 = C_TMP.emplace(C_TMP.end(), g);
+
+        // A and B are logically const; these just need to have the
+        // right alignment and "stream affinity".
+        SetSyncInfo(A1->Matrix(), stream_one);
+        SetSyncInfo(B1->Matrix(), stream_two);
+
+        A1->Resize(A.Height(), bsize);
+        B1->Resize(B.Width(), bsize);
+
+        A1->AlignWith(C);
+        B1->AlignWith(C);
+
+        // The copies of C should be initialized to zero so the
+        // accumulation is correct.
+        if (id == 0UL)
+        {
+            View(*C1, C);
+            SetSyncInfo(C1->Matrix(), stream_two);
+        }
+        else
+        {
+            SetSyncInfo(C1->Matrix(), stream_two);
+            C1->Resize(C.Height(), C.Width());
+            C1->AlignWith(C);
+
+            // Zero things out.
+            Zero(*C1);
+        }
+    }
+
+    // From this point on, we don't use the stream pool explicitly;
+    // matrices are directly queried for their stream
+    // information. This seemed less prone to error.
+
+    // Compute the rank-k updates. This is Allgather-Allgather, Compute.
+    size_t team_id = 0;
+    for (Int k = 0; k < sumDim; k += bsize)
+    {
+        // Ultimately, these are "locked views" of ranges of
+        // columns/rows of A/B, resp.
+        DistMatrix<T,MC,MR,ELEMENT,D> A1(g), B1(g);
+        Int const nb = Min(bsize, sumDim-k);
+
+        auto& AMCSTAR = A1_MC_STAR[team_id];
+        auto& BTMRSTAR = B1Trans_MR_STAR[team_id];
+
+        // Set the streams for the views so operations with them are
+        // correctly synchronized and ordered.
+        auto const& stream_one =
+            SyncInfoFromMatrix(AMCSTAR.Matrix());
+        auto const& stream_two =
+            SyncInfoFromMatrix(BTMRSTAR.Matrix());
+
+        SetSyncInfo(A1.Matrix(), stream_one);
+        SetSyncInfo(B1.Matrix(), stream_two);
+
+        // Select the block of columns/rows of A/B.
+        A1 = A(ALL,        IR(k,k+nb));
+        B1 = B(IR(k,k+nb), ALL       );
+
+        // Perform data movement (allgathers).
+        AMCSTAR = A1;
+        Transpose(B1, BTMRSTAR);
+
+        // Compute the local portion of the rank-k update. This is
+        // stored in the "team-local" storage. This assures no data
+        // race in updates to C (since C isn't being updated yet).
+        LocalGemm(NORMAL, TRANSPOSE,
+                  alpha,
+                  AMCSTAR,
+                  BTMRSTAR,
+                  TypeTraits<T>::One(), C_TMP[team_id]);
+
+        // Bookkeeping.
+        team_id = (team_id + 1) % num_stream_teams;
+    }
+
+    AddSynchronizationPoint(SyncInfoFromMatrix(C_TMP.front().LockedMatrix()),
```

Comment:
Shouldn't we synchronize `C_TMP[0]` and `C` before the computation since they use the same buffer?

Reply:
Yes, good catch.